source/Calamari.Tests/NewPipeline/PipelineCommandFixture.cs

```diff
@@ -0,0 +1,451 @@
+using Calamari.Common.Commands;
+using Calamari.Common.Features.Behaviours;
+using Calamari.Common.Features.Deployment;
+using Calamari.Common.Features.Scripts;
+using Calamari.Common.Plumbing.FileSystem;
+using Calamari.Common.Plumbing.Logging;
+using Calamari.Common.Plumbing.Pipeline;
+using Calamari.Common.Plumbing.Variables;
+using Calamari.Deployment;
+using Calamari.Testing;
+using FluentAssertions;
+using NUnit.Framework;
+
+namespace Calamari.Tests.NewPipeline
+{
+    [TestFixture]
+    public class PipelineCommandFixture
+    {
+        readonly string[] defaultScriptStages = { DeploymentStages.PreDeploy, DeploymentStages.Deploy, DeploymentStages.PostDeploy };
+
+        [Test]
+        public Task EnsureAllContainerRegistrationsAreMet()
+        {
+            return CommandTestBuilder.CreateAsync<MyCommand, MyProgram>()
+                              .Execute();
+        }
+
+        [Test]
+        public Task AssertNextBehaviourIsNotCalledWhenPreviousBehaviourSetsSkipFlag()
+        {
+            return CommandTestBuilder.CreateAsync<SkipNextCommand, MyProgram>()
+                                    .Execute();
+        }
+
+        [Test]
+        public Task AssertPipelineExtensionsAreExecuted()
+        {
+            return CommandTestBuilder.CreateAsync<MyLoggingCommand, MyProgram>()
+                                     .WithArrange(context =>
+                                                  {
+                                                      context.Variables.Add(ActionVariables.Name, "Boo");
+                                                  })
+                                    .WithAssert(result =>
+                                                {
+                                                    result.OutputVariables["ExecutionStages"].Value.Should().Be("IBeforePackageExtractionBehaviour;IAfterPackageExtractionBehaviour;IPreDeployBehaviour;IDeployBehaviour;IPostDeployBehaviour");
+                                                })
+                                    .Execute();
+        }
+
+        [Test]
+        public async Task AssertStagedPackageIsExtracted()
+        {
+            using (var tempPath = TemporaryDirectory.Create())
+            {
+                File.WriteAllText(Path.Combine(tempPath.DirectoryPath, "Hello.html"), "Hello World!");
+
+                await CommandTestBuilder.CreateAsync<MyCommand, MyProgram>()
+                                        .WithArrange(context =>
+                                                     {
+                                                         context.WithNewNugetPackage(tempPath.DirectoryPath, "MyPackage", "1.0.0");
+                                                     })
+                                        .WithAssert(result =>
+                                                    {
+                                                        File.Exists(Path.Combine(tempPath.DirectoryPath, "Hello.html")).Should().BeTrue();
```

Comment:
With all of the tests below, should we be making those assertions inside the `Deploy()` stage rather than at the end?

Reply:
Not sure I understand the difference?

Reply:
Strictly speaking these results need to be available by the time `Deploy()` is executed, so it would be more accurate to assert on that